staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go

```diff
@@ -38,8 +39,24 @@ func findServicePort(svc *v1.Service, port int32) (*v1.ServicePort, error) {
 	return nil, errors.NewServiceUnavailable(fmt.Sprintf("no service port %d found for service %q", port, svc.Name))
 }
 
-// ResourceLocation returns a URL to which one can send traffic for the specified service.
-func ResolveEndpoint(services listersv1.ServiceLister, endpoints listersv1.EndpointsLister, namespace, id string, port int32) (*url.URL, error) {
+// ResolveEndpoint returns a URL to which one can send traffic for the specified service.
+// This method accepts an optional parameter (excludedEndpoints) for specifying a list of endpoints that must be excluded.
+func ResolveEndpoint(services listersv1.ServiceLister, endpoints listersv1.EndpointsLister, namespace, id string, port int32, excludedEndpoints ...*url.URL) (*url.URL, error) {
```

Comment:
I'm surprised that exclusion is related to retries.  A second PR hiding in here for a different purpose?